### ISSUE TITLE
Support Nix in cabal-install

### ISSUE THREAD

**Issue by ttuegel** [haskell/cabal#3651]
Stack has been integrated with Nix for some time. We are actually in a better position in terms of integration because we keep persistent configuration data. Full integration would take a modest amount of additional work:
- [ ] add command-line flags and configuration options to enable or disable integration
- [ ] if Nix is enabled and the project has a `shell.nix` file, instantiate it to a `shell.drv` during `cabal configure`
- [ ] if Nix is enabled, wrap all calls to `cabal ...` in `nix-shell`: `nix-shell --run "cabal ..."`

This is slightly different than Stack's approach because it requires a `shell.nix` file, but the `cabal2nix` utility can generate those automatically from any `.cabal` file (and most Nix+Haskell users would be familiar with this tool).

**Comment by ezyang** [comment]
Not a Nix-user, but if a Nixer (I guess you?) maintains the code I don't see any problem with putting this into cabal-install.

**Comment by ttuegel** [comment]
> Not a Nix-user, but if a Nixer (I guess you?) maintains the code I don't see any problem with putting this into cabal-install.

That's the idea, to wean me away from Stack! :wink: 

For posterity, here is the zsh function I've been using to do the same thing:
```.zsh
nix-cabal () {
    CABAL=$(whence -p cabal)  # always use same cabal as current shell
    if [[ -a "$PWD/.shell.drv" ]]; then
        nix-shell --add-root "$PWD/.result" --indirect \
            "$PWD/.shell.drv" --run "$CABAL $*"
    elif [[ -a "$PWD/shell.nix" ]]; then
        nix-instantiate --add-root "$PWD/.shell.drv" --indirect \
            "$PWD/shell.nix" -A env
        nix-shell --add-root "$PWD/.result" --indirect \
            "$PWD/.shell.drv" --run "$CABAL $*" || rm "$PWD/.shell.drv"
    else
        nix-shell --add-root "$PWD/.result" --indirect --run "$CABAL $*"
    fi
}
```
I realize that probably means next to nothing to non-Nix-users; I just wanted to illustrate that the maintenance burden can be small because the logic doesn't need to be as complicated as Stack makes it.

**Comment by ryantrinkle** [comment]
@ttuegel That looks really nice!  I hope you don't mind me stealing some of that logic for my own scripts :)

**Comment by guaraqe** [comment]
This would be extremely welcome!

**Comment by ezyang** [comment]
Not a priority for 2.0, but @ttuegel if you think you can figure something out in time for 2.0 we can remilestone

**Comment by ttuegel** [comment]
@ezyang I agree it's not a priority for 2.0. I've discussed this with several Nix people, and I think it will work better with new-build anyway. Maybe we could make a milestone for 3.0? I usually milestone things as `_|_` only when nobody has any intention of working on them.

**Comment by ttuegel** [comment]
I am re-milestoning this because #3728 is blocked on it; however, I don't think either should be a release blocker.

**Comment by ttuegel** [comment]
After some conversation with @ezyang, I think that #3728 is not really blocked on this. I am going to leave it on the 2.0 milestone at low priority, though, because I have already implemented most of it :smiley:

**Comment by 23Skidoo** [comment]
@ttuegel Can this ticket be closed as fixed?

**Comment by ttuegel** [comment]
Indeed, I pushed the fix for this but completely forgot about the issue!

**Comment by ysangkok** [comment]
Seems like this was fixed in #4193
